### ISSUE TITLE
Segmentation Fault using httpchk / maxconn 

### ISSUE THREAD

**Issue by Julien75013** [haproxy/haproxy#1572]
### Detailed Description of the Problem

HAproxy doesn't start with this configuration file, it goes into a segfault state.

I deleted almost all the file, it remains only the part that allows to understand the bug.

I use php8-fpm installed on debian, with /ping enable in the conf files, answer is "pong".

### Expected Behavior

No segfault, like previous version of haproxy 2.4.4 🙂🍀

### Steps to Reproduce the Behavior

See the configuration file.

### What is your configuration?

```haproxy
global
       maxconn 5000
       daemon

fcgi-app php8-fpm
       docroot /var/www

listen the_world
#       bind 127.0.0.1:8080 tfo alpn h2,http/1.1
       mode http
       option httpchk
#       http-check send meth GET uri /ping ver HTTP/1.1 hdr Host monitoring hdr Connection close
#       http-check expect string pong
#       http-check disable-on-404
       timeout connect 1s
       timeout client 30s
       timeout server 24h
       use-fcgi-app php8-fpm
       server webserver 10.0.0.2:9000 proto fcgi check port 9000 observe layer7

```

### Output of `haproxy -vv`

```plain
HAProxy version 2.5.4-1~bpo10+1 2022/02/25 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.4.html
Running on: Linux 5.10.0-6-cloud-amd64 #1 SMP Debian 5.10.28-1 (2021-04-09) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.5.4=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
Feb 26 18:34:46 webserver kernel: [23170956.691858] haproxy[3749975]: segfault at 4000 ip 000055f987c15e4b sp 00007ffe12f837d0 error 4 in haproxy[55f987ba0000+1d6000]
Feb 26 18:34:46 webserver kernel: [23170956.694857] Code: 83 e2 10 0f 85 26 06 00 00 8b 71 04 f7 c6 00 00 00 04 0f 85 2a 06 00 00 83 e6 10 0f 84 26 15 00 00 4c 8b 71 08 4d 85 f6 74 25 <41> 80 3e 08 0f 85 5b 0d 00 00 4d 8b 76 08 4d 85 f6 74 12 41 f6 46

Feb 26 18:34:46 webserver kernel: [23170956.939935] haproxy[3749984]: segfault at a4 ip 000055e90571b36f sp 00007ffefadfba30 error 4 in haproxy[55e9056a4000+1d6000]
Feb 26 18:34:46 webserver kernel: [23170956.942412] Code: 8b 44 24 10 41 81 48 04 00 20 00 00 49 8b 80 88 00 00 00 48 89 44 24 10 e9 57 f1 ff ff 66 0f 1f 44 00 00 48 8b b1 30 fe ff ff <8b> 8e a4 00 00 00 4c 8b 4e 10 83 e1 02 4d 85 c9 0f 84 b1 04 00 00
```

### Additional Information

Case 1 :
With this configuration file, the daemon reach the maximum tries and segfault all the times immediately.

Case 2 :
When "maxconn" is commented out, it can run longer but segfault after a few moment. It's the most funny part i found, but i ignore what is the influence / impact of this parameter in this case.

For additional information :
HAProxy version 2.4.4-1~bpo10+1, it was working without issue.

**Comment by wtarreau** [comment]
Thanks!

At first I was confused because your config is missing a "bind" line. I added one and it didn't fail. I managed to get the crash by pointing the port to a local HTTP listener, such as the listener itself. This one reliably crashes it:

```
fcgi-app php8-fpm
        docroot /var/www

listen the_world
        bind :8001
        mode http
        option httpchk
        timeout connect 1s
        timeout client 30s
        timeout server 24h
        use-fcgi-app php8-fpm
        server webserver 127.0.0.1:8001 proto fcgi check
```

**Comment by wtarreau** [comment]
Forgot to attach the trace:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000004ba81b in sess_src (sess=0x0) at include/haproxy/session.h:242
242             struct connection *cli_conn = objt_conn(sess->origin);
[Current thread is 1 (Thread 0x7fe03a3c7640 (LWP 5083))]
(gdb) bt
#0  0x00000000004ba81b in sess_src (sess=0x0) at include/haproxy/session.h:242
#1  0x00000000004bac18 in si_src (si=0xdba050) at include/haproxy/stream_interface.h:577
#2  0x00000000004be8c3 in fcgi_set_default_param (fconn=0x7fe01c0262c0, fstrm=0x7fe01c0268d0, htx=0x7fe01c0269f0, sl=0x7fe01c026a20, params=0x7fe03a3a3590) at src/mux_fcgi.c:1232
#3  0x00000000004c3ff0 in fcgi_strm_send_params (fconn=0x7fe01c0262c0, fstrm=0x7fe01c0268d0, htx=0x7fe01c0269f0) at src/mux_fcgi.c:2093
#4  0x00000000004d0bc3 in fcgi_snd_buf (cs=0x7fe01c026080, buf=0xdba048, count=67, flags=0) at src/mux_fcgi.c:4071
#5  0x0000000000508abf in tcpcheck_eval_send (check=0xdb9fe8, rule=0xdb99b0) at src/tcpcheck.c:1488
#6  0x000000000050d319 in tcpcheck_main (check=0xdb9fe8) at src/tcpcheck.c:2218
#7  0x000000000060f664 in wake_srv_chk (cs=0x7fe01c026080) at src/check.c:1031
#8  0x000000000060f979 in event_srv_chk_io (t=0xdbfe50, ctx=0xdb9fe8, state=32832) at src/check.c:1081
#9  0x0000000000687f79 in run_tasks_from_lists (budgets=0x7fe03a3a44d0) at src/task.c:580
#10 0x0000000000688bd3 in process_runnable_tasks () at src/task.c:883
#11 0x0000000000631107 in run_poll_loop () at src/haproxy.c:2699
#12 0x00000000006315ce in run_thread_poll_loop (data=0x8e5d80 <ha_thread_info+256>) at src/haproxy.c:2870
#13 0x00007fe04179be45 in start_thread () from /lib64/libpthread.so.0
#14 0x00007fe04117e4af in clone () from /lib64/libc.so.6
```

**Comment by wtarreau** [comment]
OK I could bisect it to this commit:
```
commit 568008d19917d1426cb6228df37d9f3c189b4968
Author: Christopher Faulet <cfaulet@haproxy.com>
Date:   Mon Oct 25 07:56:51 2021 +0200

    MINOR: mux-fcgi: Rely on client addresses at stream level to set default params
    
    Client source and destination addresses at stream level are now used to emit
    SERVER_NAME/SERVER_PORT and REMOTE_ADDR/REMOTE_PORT parameters. For now,
    stream-interface addresses are never set. So, thanks to the fallback
    mechanism, no changes are expected with this patch. But its purpose is to
    rely on addresses at the stream level, when set, instead of those at the
    connection level.

 src/mux_fcgi.c | 18 ++++++++++--------
```

It was introduced as part of a larger series needed to distinguish the session's source and the stream's source addresses. No sure why there's no session here, maybe that's related to checks, but then I don't see why it doesn't crash on H1/H2. CCing @capflam.

**Comment by Julien75013** [comment]
@wtarreau thanks to you

It's true reusing the "same port" you reported.

However I updated the initial issue (actually the "bind" line was not necessary for my segfault), as it may be another crash, not sure. I will try to compile / gdb if you cannot try to crash with my configuration. Let me know.

**Comment by capflam** [comment]
Indeed, there is no stream/stream-interface for health-check. @wtarreau, on H1/H2, the client source and destination addresses are never retrieved. Thus there is no reason to have a crash. For the FCGi multiplexer, we must provide them. Here is a quick patch for HAProxy-2.5:
```diff
diff --git a/src/mux_fcgi.c b/src/mux_fcgi.c
index ba3a54617..01f0cb7fd 100644
--- a/src/mux_fcgi.c
+++ b/src/mux_fcgi.c
@@ -1229,8 +1229,8 @@ static int fcgi_set_default_param(struct fcgi_conn *fconn, struct fcgi_strm *fst
                                  struct fcgi_strm_params *params)
 {
        struct connection *cli_conn = objt_conn(fstrm->sess->origin);
-       const struct sockaddr_storage *src = si_src(si_opposite(fstrm->cs->data));
-       const struct sockaddr_storage *dst = si_dst(si_opposite(fstrm->cs->data));
+       const struct sockaddr_storage *src = (objt_check(fstrm->sess->origin) ? conn_src(fconn->conn) : si_src(si_opposite(fstrm->cs->data)));
+       const struct sockaddr_storage *dst = (objt_check(fstrm->sess->origin) ? conn_dst(fconn->conn) : si_dst(si_opposite(fstrm->cs->data)));
        struct ist p;
 
        if (!sl)
```

For health-checks, we rely on the server connection because there is no client connection.

**Comment by capflam** [comment]
The fix was pushed in 2.6-DEV. It will be backported ASAP. Thanks !
